pkg/process/monitor/process_monitor.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/gopsutil/process"
+)
+
+const (
+	processMonitorMaxEvents = 2048
+)
+
+const (
+	STOPPED = iota
+	RUNNING
+)
+
+var (
+	processMonitor *ProcessMonitor
+)
+
+type ProcessMonitor struct {
+	m  sync.Mutex
+	wg sync.WaitGroup
+
+	state int
+
+	events chan netlink.ProcEvent
+	done   chan struct{}
+	errors chan error
+
+	procEventCallbacks map[ProcessEventType][]*ProcessCallback
+	callbackRunner     chan func()
+	callbackRunnerDone chan struct{}
+}
+
+type ProcessEventType int
+
+const (
+	EXEC ProcessEventType = iota
+	FORK
+	EXIT
+)
+
+type ProcessMetadataField int
+
+const (
+	ANY ProcessMetadataField = iota
+	NAME
+	MAPFILE
+)
+
+type ProcessCallback struct {
+	Event    ProcessEventType
+	Metadata ProcessMetadataField
+	Regex    *regexp.Regexp
+	Callback func(pid uint32)
+}
+
+func (pm *ProcessMonitor) enqueueCallback(callback *ProcessCallback, pid uint32) {
+	pm.callbackRunner <- func() { callback.Callback(pid) }
+}
+
+func (pm *ProcessMonitor) InitWithAllCurrentProcess() {
+	fn := func(pid int) error {
+		pm.m.Lock()
+		for _, c := range pm.procEventCallbacks[EXEC] {
+			pm.evalEXECCallback(c, uint32(pid))
+		}
+		pm.m.Unlock()
+		return nil
+	}
+
+	if err := util.WithAllProcs(util.HostProc(), fn); err != nil {
+		log.Errorf("process monitor init, scanning all process failed %s", err)
+	}
+
+	pm.state = RUNNING
+}
+
+func (p *ProcessMonitor) evalEXECCallback(c *ProcessCallback, pid uint32) {
+	if c.Metadata == ANY {
+		p.enqueueCallback(c, pid)
+		return
+	}
+
+	proc, err := process.NewProcess(int32(pid))
+	if err != nil {
+		log.Errorf("process %d parsing failed %s", pid, err)
+		return
+	}
+	switch c.Metadata {
+	case NAME:
+		pname, err := proc.Name()
+		if err != nil {
+			log.Errorf("process %d name parsing failed %s", pid, err)
+			return
+		}
+		if c.Regex.MatchString(pname) {
+			p.enqueueCallback(c, pid)
+		}
+	case MAPFILE:
+		mmaps, err := proc.MemoryMaps(true)
+		if err != nil {
+			log.Errorf("process %d/smaps parsing failed %s", pid, err)
+			return
+		}
+		if mmaps == nil {
+			return
+		}
+		for _, mmap := range *mmaps {
+			if c.Regex.MatchString(mmap.Path) {
+				p.enqueueCallback(c, pid)
+			}
+		}
+	}
+}
+
+func (pm *ProcessMonitor) Subscribe(callback *ProcessCallback) (UnSubscribe func()) {
+	pm.m.Lock()
+	defer pm.m.Unlock()
+
+	pm.procEventCallbacks[callback.Event] = append(pm.procEventCallbacks[callback.Event], callback)
+
+	return func() {
+		pm.m.Lock()
+		defer pm.m.Unlock()
+
+		for i, c := range pm.procEventCallbacks[callback.Event] {
+			if c == callback {
+				l := len(pm.procEventCallbacks[callback.Event])
+				pm.procEventCallbacks[callback.Event][i] = pm.procEventCallbacks[callback.Event][l-1]
+				pm.procEventCallbacks[callback.Event] = pm.procEventCallbacks[callback.Event][:l-1]
+			}
+		}
+	}
+}
+
+func (pm *ProcessMonitor) Stop() {
+	close(pm.callbackRunnerDone)
+	close(pm.done)
+	pm.wg.Wait()
+	processMonitor = nil
```

Comment:
1. why `processMotitor` and not `pm`?
2. the entire method should be wrapped with `m.Lock`

Reply:
1/ processMonitor is used as cache value and this invalidate it
2/ thanks I added rwlock, even if there are no race there (1 thread / 1 path for the init and cleanup)

Reply:
I think we can omit the rwlock if thread safety is not required. just explicitly mention next to the global var, that it is a singleton non-thread safe implementation.

if for some reason we do want to keep thread-safe behavior, let's use `once.Do` and  we shouldn't reset the global pointer from one of the methods of the instance

Reply:
I added the rwlock yesterday, something was out of sync .... `once.Do()` will not guarantee right initialization of 2 concurrent threads. ex: one will initialize (once.Do()) the other(s) return nil

Reply:
actually Once.Do takes care of exactly this situation and the subsequent calls are [waiting](https://go.dev/src/sync/once.go?s=1137:1164#L43) until the first call is completes.

this is the standard way to [implement](https://medium.easyread.co/just-call-your-code-only-once-256f69ed39a8) the thread safe singleton pattern in go

Reply:
yep I will move to Once.Do
TIL from the doc (at the end) `Because no call to Do returns until the one call to f returns, if f causes Do to be called, it will deadlock.`